### PR TITLE
Rework constructors for LinearAlgebra

### PR DESCRIPTION
**Pull request by timholy**
This started life as part of #24899, but I couldn't help myself from making other changes and so I'm splitting it out here. Note that #24899 is not complete without this---it assumed that certain types had been "pre-checked" for validity---so even if we don't accept some of this we'll need to add the calls to `has_offset_axes` to the inner constructors.

Aside from enforcing 1-indexing, these allow one to coerce some of the field types via construction (without requiring that the inputs already have those types). Overall I believe these make the LinearAlgebra constructors more flexible and consistent. Note that certain type signatures have changed a bit, e.g., `Hessenberg{T,S<:AbstractMatrix{T}}` instead of `Hessenberg{T,S<:AbstractMatrix}`. In one place this actually led to a need to change the result of one of our tests; I'll annotate that line directly.

### CODE REVIEW

File: stdlib/LinearAlgebra/test/tridiag.jl
Comment (timholy) [inline stdlib/LinearAlgebra/test/tridiag.jl]: I found the old result here surprising; for example, when passed a `SymTridiagonal{Float64,Vector{Float64}}` this conversion created a `LDLt{Float32,SymTridiagonal{Float64,Vector{Float64}}}`, meaning it claimed Float32 precision but internally it was stored with Float64 precision. By changing the type declaration from `LDLt{T,S<:AbstractMatrix}` to `LDLt{T,S<:AbstractMatrix{T}}` I was forced to change this outcome to `LDLt{Float32,SymTridiagonal{Float32,Vector{Float32}}}`. Was the previous outcome actually desired? Or was it essentially an oversight?

Comment (andreasnoack) [reply stdlib/LinearAlgebra/test/tridiag.jl]: I'm pretty sure that must have been an oversight.

Comment (andreasnoack) [reply stdlib/LinearAlgebra/test/tridiag.jl]: I think these definitions and tests predates the time when `LDLt{T,S<:AbstractMatrix{T}}` was possible.

**Review by andreasnoack — COMMENTED** [review]
Thanks for cleaning up the constructors. They were indeed a bit inconsistent.

I don't understand why factorizations can't have offset axes.

File: stdlib/LinearAlgebra/src/bidiag.jl
Comment (timholy) [reply stdlib/LinearAlgebra/src/bidiag.jl]: I was guarding against poorly-written `convert` methods that don't actually do what you expect them to do. In at least some cases I was worried that if they failed to convert, one could trigger a `StackOverflowError`. By adding the typeassert it guarantees that the error will be easily understood.

File: stdlib/LinearAlgebra/src/lq.jl
Comment (timholy) [reply stdlib/LinearAlgebra/src/lq.jl]: I may have been too conservative and just added checks to all the special matrix types. In general I was concerned about methods like [this one](https://github.com/JuliaLang/julia/blob/f104ea4ec352519de76e52bf65ea7b3ed2dc6155/stdlib/LinearAlgebra/src/lq.jl#L261-L268) which check bounds in a way that ignores the starting index.

File: stdlib/LinearAlgebra/test/tridiag.jl
Comment (timholy) [reply stdlib/LinearAlgebra/test/tridiag.jl]: I'm not surprised. I have a ton of constructors that need fixing littered through the package landscape and my lab's own internal code.

File: stdlib/LinearAlgebra/src/lq.jl
Comment (timholy) [reply stdlib/LinearAlgebra/src/lq.jl]: That was pretty much my plan too, :+1:

**Comment by timholy** [comment]
Tests seem OK to me. I'll leave it up to others to decide if/when to merge this, since I don't have anything I'm trying to finish that depends on this. If necessary feel free to assume ownership over this.

**Review by andreasnoack — APPROVED** [review]

**Comment by andreasnoack** [comment]
This shouldn't cause any trouble so I'll merge tomorrow unless anybody complains.

**Pull request merged**
